src/uopool/mod.rs

```diff
@@ -1,36 +1,65 @@
 use crate::{
     types::user_operation::{UserOperation, UserOperationHash},
-    uopool::{server::uopool_server::uo_pool_server::UoPoolServer, services::UoPoolService},
+    uopool::{
+        memory::MemoryMempool, server::uopool_server::uo_pool_server::UoPoolServer,
+        services::UoPoolService,
+    },
 };
 use anyhow::Result;
+use async_trait::async_trait;
 use clap::Parser;
 use educe::Educe;
-use ethers::types::Address;
+use ethers::{
+    abi::AbiEncode,
+    types::{Address, H256, U256},
+    utils::keccak256,
+};
 use jsonrpsee::tracing::info;
-use parking_lot::RwLock;
-use std::{collections::HashMap, net::SocketAddr, sync::Arc, time::Duration};
+use std::{fmt::Debug, net::SocketAddr, sync::Arc, time::Duration};
 
+pub mod memory;
 pub mod server;
 pub mod services;
 
-#[derive(Educe)]
-#[educe(Debug)]
-pub struct UserOperationPool {
-    pub pool: Arc<RwLock<HashMap<UserOperationHash, UserOperation>>>,
-}
+pub type MempoolId = H256;
 
-impl UserOperationPool {
-    pub fn new() -> Self {
-        Self {
-            pool: Default::default(),
-        }
+#[async_trait]
+pub trait Mempool: Debug + Send + Sync + 'static {
+    fn id(entry_point: Address, chain_id: U256) -> MempoolId {
+        H256::from_slice(keccak256([entry_point.encode(), chain_id.encode()].concat()).as_slice())
     }
+    fn add(
```

Comment:
I am curious. Why not make these methods in `async` way? In that way, we could support cocurrent.

Reply:
Hmm, missed that.